File: Client/iOS/Component/ToolKit/Common/ComponentProtocol/RTCTokenProtocol/RTCTokenProtocol.m
// Copyright (c) 2023 BytePlus Pte. Ltd.
// SPDX-License-Identifier: Apache-2.0
//

#import "RTCTokenProtocol.h"

@implementation RTCTokenProtocol

- (void)tokenWithAppID:(NSString *)appId
                appKey:(NSString *)appKey
                roomId:(NSString *)roomId
                   uid:(NSString *)userId
                 block:(nonnull void (^)(NSString * _Nonnull))block {
    NSObject<RTCTokenDelegate> *delegate = [[NSClassFromString(@"TokenGenerator") alloc] init];

    if ([delegate respondsToSelector:@selector(protocol:tokenWithAppID:appKey:roomId:uid:block:)]) {
        [delegate protocol:self
            tokenWithAppID:appId
                    appKey:appKey
                    roomId:roomId
                       uid:userId
                     block:block];
    }
}

- (void)getRTMPAddr:(NSString *)taskID
              block:(void (^)(NSString *addr))block {
    NSObject<RTCTokenDelegate> *delegate = [[NSClassFromString(@"TokenGenerator") alloc] init];

    if ([delegate respondsToSelector:@selector(protocol:getRTMPAddr:block:)]) {
        [delegate protocol:self getRTMPAddr:taskID block:block];
    }
}



@end

File: Client/iOS/Component/RTCAPIExample/RTCAPIExample/Common/RTCTokenProtocol.swift
// Copyright (c) 2024 BytePlus Pte. Ltd.
// SPDX-License-Identifier: Apache-2.0
//

import Foundation

@objc
public protocol RTCTokenProtocol: NSObjectProtocol {
    static func rtcAppId() -> String
    static func generateToken(roomId: String, userId: String, completion: @escaping (String?) -> Void)
}

func rtcAppId() -> String {
    guard let clazz = NSClassFromString("TokenGenerator") as? RTCTokenProtocol.Type else {
        return ""
    }

    return clazz.rtcAppId()
}

// To obtain the token, it is recommended to obtain it from the server
func generateToken(roomId: String, userId: String, completion: @escaping (String?) -> Void) {
    guard let clazz = NSClassFromString("TokenGenerator") as? RTCTokenProtocol.Type else {
        completion("")
        return
    }

    return clazz.generateToken(roomId: roomId, userId: userId, completion: completion)
}
